### PR TITLE
Api 835 r4 medication model

### PR DESCRIPTION
**Pull request by taylorrassmannlibertyits**
First pull request! Please tell me all the things I did wrong / right :)

### CODE REVIEW

File: us-core-r4-examples/src/main/java/gov/va/api/health/uscorer4/api/swaggerexamples/SwaggerMedication.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    return Medication.Bundle.builder()
+        .resourceType("Bundle")
+        .type(AbstractBundle.BundleType.searchset)
+        .total(1)
+        .link(
+            asList(
+                BundleLink.builder()
+                    .relation(BundleLink.LinkRelation.self)
+                    .url(
+                        "https://sandbox-api.va.gov/services/argonaut/v0/r4/Medication?"
```
Comment (jhulbert-va) [inline us-core-r4-examples/src/main/java/gov/va/api/health/uscorer4/api/swaggerexamples/SwaggerMedication.java:52]: `services/fhir/v0/r4`

File: us-core-r4/src/test/java/gov/va/api/health/uscorer4/api/samples/SampleMedications.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        .itemReference(reference())
+        .build();
+  }
+
+  public Medication medication() {
+    return Medication.builder()
+        .id("1234")
+        .meta(meta())
+        .implicitRules("http://GoodnightRules.com")
+        .language("Goodnight Language")
```
Comment (bschofield-va) [inline us-core-r4/src/test/java/gov/va/api/health/uscorer4/api/samples/SampleMedications.java:37]: Good night light, and the red balloon

File: us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+  @Data
+  @NoArgsConstructor
+  @EqualsAndHashCode(callSuper = true)
+  @JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
+  @JsonDeserialize(builder = Medication.Bundle.BundleBuilder.class)
+  @Schema(
+      name = "MedicationBundle",
+      example =
+          "${r4.medicationBundle:gov.va.api.health."
+              + "r4.api.swaggerexamples.SwaggerMedication#medicationBundle}")
```
Comment (jhulbert-va) [inline us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java:107]: `uscorer4`

File: us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+  public static class Ingredient implements BackboneElement {
+    String id;
+
+    @Valid List<Extension> extension;
+
+    @Valid List<Extension> modifierExtension;
+
+    @Valid CodeableConcept itemCodeableConcept;
+
+    @Valid @NotNull Reference itemReference;
```
Comment (jhulbert-va) [inline us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java:182]: Doesnt need `@NotNull`. Exactly one of will enforce that.

File: us-core-r4-examples/src/main/java/gov/va/api/health/uscorer4/api/swaggerexamples/SwaggerMedication.java (hunk 1 above)
Comment (jblefkove-va) [inline us-core-r4-examples/src/main/java/gov/va/api/health/uscorer4/api/swaggerexamples/SwaggerMedication.java]: This url doesn't match the others

File: us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java (hunk 4 above)
Comment (bschofield-va) [reply us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java:182]: In fact, `@NotNull` is incompatible with `@ExactlyOne`

File: us-core-r4-examples/src/main/java/gov/va/api/health/uscorer4/api/swaggerexamples/SwaggerMedication.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  /**
+   * An example Medication.
+   *
+   * @return an example Medication.
+   */
+  public static Medication medication() {
+    return Medication.builder()
+        .resourceType("Medication")
+        .id("I2-U4FPJS3E633MAJQBCAA2KAB5BQ000000")
+        .status(Medication.Status.active)
```
Comment (jblefkove-va) [inline us-core-r4-examples/src/main/java/gov/va/api/health/uscorer4/api/swaggerexamples/SwaggerMedication.java:24]: This example seems a bit sparse? Take a look at these R4 examples: http://hl7.org/fhir/R4/medication-examples.html

File: us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/MedicationApi.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+          "https://build.fhir.org/ig/HL7/US-Core-R4/StructureDefinition-us-core-medication.html",
+      tags = {"Medication"})
+  @GET
+  @Path("Medication/{id}")
+  @ApiResponse(
+      responseCode = "200",
+      description = "Record found",
+      content =
+          @Content(
+              mediaType = "application/json+fhir",
```
Comment (bschofield-va) [inline us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/MedicationApi.java:28]: should be `application/fhir+json`
http://hl7.org/fhir/R4/http.html#mime-type

File: us-core-r4-examples/src/main/java/gov/va/api/health/uscorer4/api/swaggerexamples/SwaggerMedication.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+                    .relation(BundleLink.LinkRelation.self)
+                    .url(
+                        "https://sandbox-api.va.gov/services/argonaut/v0/r4/Medication?"
+                            + "patient=1017283148V813263&page=1&_count=15")
+                    .build(),
+                BundleLink.builder()
+                    .relation(BundleLink.LinkRelation.first)
+                    .url(
+                        "https://sandbox-api.va.gov/services/fhir/v0/r4/Medication?"
+                            + "patient=1017283148V813263&page=1&_count=15")
```
Comment (jblefkove-va) [inline us-core-r4-examples/src/main/java/gov/va/api/health/uscorer4/api/swaggerexamples/SwaggerMedication.java:60]: Medication doesn't support search by patient. How about `_id=I2-U4FPJS3E633MAJQBCAA2KAB5BQ000000` instead

File: us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/MedicationApi.java (hunk 6 above)
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/MedicationApi.java:28]: These should be `fhir+json`

File: us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+
+  @Valid Reference manufacturer;
+
+  @Valid CodeableConcept form;
+
+  @Valid Ratio amount;
+
+  @Valid List<Ingredient> ingredient;
+
+  @Valid List<Batch> batch;
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java:90]: Only one batch

File: us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java (hunk 3 above)
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java]: package should be `uscorer4`

File: us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java (hunk 4 above)
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java]: This field can be null... the `ExactlyOneOf` validation above will handle this

File: us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+  @Schema(name = "MedicationIngredient")
+  @AllArgsConstructor
+  @NoArgsConstructor(access = AccessLevel.PRIVATE)
+  @JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
+  @ExactlyOneOfs({
+    @ExactlyOneOf(
+        fields = {"itemCodeableConcept", "itemReference"},
+        message = "Exactly one field for item... itemCodeableConcept | itemReference")
+  })
+  public static class Ingredient implements BackboneElement {
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java:174]: Missing fields `isActive` and `strength`

File: us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+    String id;
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java:175]: `@Pattern(regexp = Fhir.ID)`

File: us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+  }
+
+  @Data
+  @Builder
+  @Schema(name = "MedicationBatch")
+  @AllArgsConstructor
+  @NoArgsConstructor(access = AccessLevel.PRIVATE)
+  @JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
+  public static class Batch implements BackboneElement {
+    String id;
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java:193]: `@Pattern(regexp = Fhir.ID)`

File: us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java (hunk 9 above)
Comment (bschofield-va) [inline us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java:174]: Missing `isActive` and `strength`

File: us-core-r4-examples/src/main/java/gov/va/api/health/uscorer4/api/swaggerexamples/SwaggerMedication.java (hunk 7 above)
Comment (taylorrassmannlibertyits) [reply us-core-r4-examples/src/main/java/gov/va/api/health/uscorer4/api/swaggerexamples/SwaggerMedication.java:60]: Cool, I didn't think it did, this makes more sense now. Updated in next commit.

File: us-core-r4-examples/src/main/java/gov/va/api/health/uscorer4/api/swaggerexamples/SwaggerMedication.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+                .build())
+        .ingredient(
+            asList(
+                Medication.Ingredient.builder()
+                    .itemCodeableConcept(
+                        CodeableConcept.builder()
+                            .text("Timolol Maleate (substance)")
+                            .coding(
+                                List.of(
+                                    Coding.builder().system("SNOMED CT").code("75359005").build()))
```
Comment (jblefkove-va) [inline us-core-r4-examples/src/main/java/gov/va/api/health/uscorer4/api/swaggerexamples/SwaggerMedication.java:49]: System should be a full url like `http://snomed.info/sct`. (Ditto line 39)

File: us-core-r4-examples/src/main/java/gov/va/api/health/uscorer4/api/swaggerexamples/SwaggerMedication.java [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+                            .build())
+                    .build()))
+        .batch(Medication.Batch.builder().lotNumber("9494788").expirationDate("22/05/2017").build())
```
Comment (jblefkove-va) [inline us-core-r4-examples/src/main/java/gov/va/api/health/uscorer4/api/swaggerexamples/SwaggerMedication.java:52]: Date format should be YYYY-MM-DD

File: us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+
+  @Data
+  @NoArgsConstructor
+  @EqualsAndHashCode(callSuper = true)
+  @JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
+  @JsonDeserialize(builder = Medication.Bundle.BundleBuilder.class)
+  @Schema(
+      name = "MedicationBundle",
+      example =
+          "${r4.medicationBundle:gov.va.api.health."
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/uscorer4/api/resources/Medication.java:106]: `uscorer4.medicationBundle`

**Pull request merged**